tests/test_runner.py

```diff
@@ -0,0 +1,304 @@
+        rr.run(model=classification_model, test_dataset=test_dataset)
+    assert len(warning_records) == 2
+    # Warning from build function
+    first_warning = warning_records[0].message.args[0]
+    assert "jupyter-book build" in first_warning
+    assert "did not succeed" in first_warning
+    assert str(rr.jb_build_log) in first_warning
+    # Warning from run function
+    second_warning = warning_records[1].message.args[0]
+    assert "expected report main page" in second_warning
+    assert "error generating" in second_warning
+    assert str(rr.jb_build_log) in second_warning
+
+    # Build artifacts mention error
+    assert rr._jb_build_result.returncode > 0
+    with open(rr.jb_build_log) as f:
+        build_log = f.read()
+    assert build_log.startswith("Running Jupyter-Book")
+    assert "Finished generating HTML" not in build_log
+    assert "error in building" in build_log.lower()
+    assert "RuntimeError" in build_log
+
+    # Report was not produced
+    assert not rr.report_main_page.exists()
+    assert not rr._linked_main_page.exists()
+    with pytest.raises(AttributeError):
+        rr.report_html
+    with pytest.raises(AttributeError):
+        rr.open()
+
+
+def test_run_report_error_config(tmp_path, classification_model, test_dataset):
+    # Passing in config file to run() overrides default config.
+    # Fake config causes error.
+    out_path_run = tmp_path / "test_run"
+    exec_path_run = tmp_path / "test_exec"
+    fake_config_file = tmp_path / "fake_config.yaml"
+    with open(fake_config_file, "w") as f:
+        yaml.dump({"fake_option": 123}, f)
+
+    rr = ReportRunner(output_path=out_path_run, execution_path=exec_path_run)
+    rr.run(
+        model=classification_model,
+        test_dataset=test_dataset,
+        config_filepath=fake_config_file,
+    )
+
+    # jupyter-book build job succeeded even though notebooks didn't
+    assert rr._jb_build_result.returncode == 0
+    with open(rr.jb_build_log) as f:
+        build_log = f.read()
+    assert build_log.startswith("Running Jupyter-Book")
+    assert "Finished generating HTML" in build_log
+    # Log file mentions failure
+    assert "failed" in build_log.lower()
+
+    # Report was created
+    assert rr.report_main_page.exists()
+    # Execution dir and output dir contains relevant files.
+    assert len(list(rr.execution_path.glob("_context_store*"))) > 0
+    with open(rr.execution_path / "_toc.yml") as f:
+        toc = [x.strip() for x in f.readlines()]
+        notebooks = [x[8:] for x in toc if x.startswith("- file: ")]
+    for nb in notebooks:
+        assert (rr.execution_path / f"{nb}.ipynb").exists()
+        assert (rr.report_main_page.parent / f"{nb}.html").exists()
+    # Error message shows in notebooks
+    with open(rr.report_main_page.parent / "conditional_metric.html") as f:
+        eval_html = f.read()
+    assert "Error" in eval_html
+    assert "Traceback" in eval_html
```

Comment:
Very thorough runner test coverage!

Reply:
:) good to cover our bases at the entrypoint